pkg/volume/portworx/portworx_util.go

```diff
@@ -0,0 +1,177 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package portworx
+
+import (
+	//"strings"
+
+	"github.com/golang/glog"
+	osdapi "github.com/libopenstorage/openstorage/api"
+	osdclient "github.com/libopenstorage/openstorage/api/client"
+	volumeclient "github.com/libopenstorage/openstorage/api/client/volume"
+	osdspec "github.com/libopenstorage/openstorage/api/spec"
+	osdvolume "github.com/libopenstorage/openstorage/volume"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	osdMgmtPort      = "9001"
+	osdDriverVersion = "v1"
+	pxdDriverName    = "pxd"
+	pwxSockName      = "pwx"
+)
+
+type PortworxVolumeUtil struct {
+	portworxClient *osdclient.Client
+}
+
+// CreateVolume creates a Portworx volume.
+func (util *PortworxVolumeUtil) CreateVolume(p *portworxVolumeProvisioner) (string, int, map[string]string, error) {
+	hostName := p.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return "", 0, nil, err
+	}
+
+	capacity := p.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	glog.V(2).Infof("Capacity provided by k8s: ", capacity)
+	// Portworx Volumes are specified in GB
+	requestGB := int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	var labels map[string]string
+	if p.options.CloudTags == nil {
+		labels = make(map[string]string)
+	} else {
+		labels = *p.options.CloudTags
+	}
+
+	specHandler := osdspec.NewSpecHandler()
+	spec, err := specHandler.SpecFromOpts(p.options.Parameters)
+	if err != nil {
+		return "", 0, nil, err
+	}
+	spec.Size = uint64(requestGB * 1024 * 1024 * 1024)
+	source := osdapi.Source{}
+	locator := osdapi.VolumeLocator{
+		Name:         p.options.PVName,
+		VolumeLabels: labels,
+	}
+	volumeID, err := client.Create(&locator, &source, spec)
+	if err != nil {
+		glog.V(2).Infof("Error creating Portworx Volume : %v", err)
+	}
+	return volumeID, requestGB, nil, err
+}
+
+// DeleteVolume deletes a Portworx volume
+func (util *PortworxVolumeUtil) DeleteVolume(d *portworxVolumeDeleter) error {
+	hostName := d.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Delete(d.volumeID)
+	if err != nil {
+		glog.V(2).Infof("Error deleting Portworx Volume (%v): %v", d.volName, err)
+		return err
+	}
+	return nil
+}
+
+// AttachVolume attaches a Portworx Volume
+func (util *PortworxVolumeUtil) AttachVolume(m *portworxVolumeMounter) (string, error) {
+	hostName := m.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return "", err
+	}
+
+	devicePath, err := client.Attach(m.volName)
+	if err != nil {
+		if err == osdvolume.ErrVolAttachedOnRemoteNode {
+			// Volume is already attached to node.
+			glog.V(2).Infof("Error attaching Portworx Volume (%v): "+
+				"Volume is already attached to another node.", m.volName)
+			return "", err
+		}
```

Comment:
generalize to just printing and returning the error since there is no specific action for that error